### PR TITLE
Split "self-service" flag in UAA zone config to independently

### PR DESCRIPTION
**Pull request by babamishra9**
toggle "Reset Password" and "Register User" functionality
- If "Reset Password" flag is enabled, user can see the
  "Reset Password" link, and navigate to "Reset Password" screen.
  If "Reset Password" flag is disabled, "Reset Password" link
  is not visible.Navigate to "Reset Password" ("/forgot_password")
  screen gives error screen with messsage "Self service reset
  password is disabled." and rest api will return 404 http status.
- The same is applicable for ""Register User" (“/create_account”)
  flag."

### CODE REVIEW

File: model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java [hunk 1]
```diff
@@ -98,16 +98,26 @@ public Logout setWhitelist(List<String> whitelist) {
     }
 
     public static class SelfService {
-        private boolean selfServiceLinksEnabled = true;
```
Comment (irinaepshteyn) [inline model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java:101]: selfServiceLinksEnabled is present in the configuration of the existing UAA zones (zone config is stored in the database as text blob).  How are we going to support backward compatibility?

Whenever zone configuration is read from the database, its selfServiceLinksEnabled and links should be parsed into assigning selfServiceResetPasswordEnabled and selfServiceCreateAccountEnabled.  Please review https://devcloud.swcoe.ge.com/devspace/display/RMOCM/Enable+self-service+links+for+UAA+instance to see different configurations.

We haven't discussed backward compatibility in the story.  So you current implementation and the current story could be about new zones and support for existing zones could be handled as another story.  For now selfServiceLinksEnabled can not be deleted.  I am not sure it should be kept here in this class, but somewhere where config is read from the database.

New zones are:
- provisioned by Service Broker "cf cs ..." - selfServiceLinksEnabled flag and links are set to default
- provisioned ad-hoc, for example, during acceptance test setup "POST /identity-zones" - selfServiceLinksEnabled flag and links are set to default unless explicitly specified in the request.

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java [hunk 2]
```diff
@@ -46,8 +46,9 @@ public String activationEmail(Model model,
                                   @RequestParam(value = "client_id", required = false) String clientId,
                                   @RequestParam(value = "redirect_uri", required = false) String redirectUri,
                                   HttpServletResponse response) {
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceCreateAccountEnabled()) {
```
Comment (irinaepshteyn) [inline server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java:49]: Create a variable for IdentityZoneHolder.get().getConfig().getLinks().getSelfService() for easier reading

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java [hunk 3]
```diff
@@ -64,8 +64,9 @@ public String forgotPasswordPage(Model model,
                                      @RequestParam(required = false, value = "client_id") String clientId,
                                      @RequestParam(required = false, value = "redirect_uri") String redirectUri,
                                      HttpServletResponse response) {
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceResetPasswordEnabled()) {
```
Comment (irinaepshteyn) [inline server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java:67]: Create a variable for IdentityZoneHolder.get().getConfig().getLinks().getSelfService() for easier reading

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
@babamishra9 , could you address my initial comments while I continue reviewing.

File: model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java (hunk 1 above)
Comment (babamishra9) [reply model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java:101]: @irinaepshteyn , As discussed, we will have another story for support backward compatibility. 
For new zones, split "self-service" flag in UAA zone config has been implemented.

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java (hunk 2 above)
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java]: Done

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java (hunk 3 above)
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java]: Done.

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
@babamishra9 , I complete my review.  Please address my comments.

File: server/src/test/java/org/cloudfoundry/identity/uaa/login/ResetPasswordControllerTest.java [hunk 4]
```diff
@@ -104,17 +104,17 @@ void testForgotPasswordPage() throws Exception {
     }
 
     @Test
-    void testForgotPasswordWithSelfServiceDisabled() throws Exception {
+    void testForgotPasswordWithSelfServiceResetPasswordEnabled() throws Exception {
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/test/java/org/cloudfoundry/identity/uaa/login/ResetPasswordControllerTest.java]: Done

File: server/src/test/java/org/cloudfoundry/identity/uaa/login/ResetPasswordControllerTest.java [hunk 5]
```diff
@@ -207,9 +207,9 @@ void forgotPassword_SuccessfulInOtherZone() throws Exception {
     }
 
     @Test
-    void forgotPasswordPostWithSelfServiceDisabled() throws Exception {
+    void forgotPasswordPostWithSelfServiceResetPasswordEnabled() throws Exception {
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/test/java/org/cloudfoundry/identity/uaa/login/ResetPasswordControllerTest.java]: Done

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java [hunk 6]
```diff
@@ -46,8 +47,10 @@ public String activationEmail(Model model,
                                   @RequestParam(value = "client_id", required = false) String clientId,
                                   @RequestParam(value = "redirect_uri", required = false) String redirectUri,
                                   HttpServletResponse response) {
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        SelfService selfService = IdentityZoneHolder.get().getConfig().getLinks().getSelfService();
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java]: Done

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java [hunk 7]
```diff
@@ -64,8 +65,10 @@ public String forgotPasswordPage(Model model,
                                      @RequestParam(required = false, value = "client_id") String clientId,
                                      @RequestParam(required = false, value = "redirect_uri") String redirectUri,
                                      HttpServletResponse response) {
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        SelfService selfService = IdentityZoneHolder.get().getConfig().getLinks().getSelfService();
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java]: Done

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java [hunk 8]
```diff
@@ -75,8 +78,10 @@ public String forgotPasswordPage(Model model,
     @RequestMapping(value = "/forgot_password.do", method = RequestMethod.POST)
     public String forgotPassword(Model model, @RequestParam("username") String username, @RequestParam(value = "client_id", defaultValue = "") String clientId,
                                  @RequestParam(value = "redirect_uri", defaultValue = "") String redirectUri, HttpServletResponse response) {
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        SelfService selfService = IdentityZoneHolder.get().getConfig().getLinks().getSelfService();
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java]: Done

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java [hunk 9]
```diff
@@ -69,8 +72,10 @@ public String sendActivationEmail(Model model, HttpServletResponse response,
         if (zoneBranding != null && zoneBranding.getConsent() != null && !doesUserConsent) {
             return handleUnprocessableEntity(model, response, "error_message_code", "missing_consent");
         }
-        if (!IdentityZoneHolder.get().getConfig().getLinks().getSelfService().isSelfServiceLinksEnabled()) {
-            return handleSelfServiceDisabled(model, response, "error_message_code", "self_service_disabled");
+        SelfService selfService = IdentityZoneHolder.get().getConfig().getLinks().getSelfService();
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/AccountsController.java]: Done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java [hunk 10]
```diff
@@ -1208,7 +1209,7 @@ void testAccessConfirmationPage(
 
     @Test
     void testSignupsAndResetPasswordEnabled() throws Exception {
-        MockMvcUtils.setSelfServiceLinksEnabled(webApplicationContext, IdentityZone.getUaaZoneId(), true);
+        MockMvcUtils.setSelfServiceResetPasswordEnabled(webApplicationContext, IdentityZone.getUaaZoneId(), true);
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java:1212]: Done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java [hunk 11]
```diff
@@ -2665,10 +2668,12 @@ void emailPageIdpDiscoveryEnabled_SelfServiceLinksDisabled(
     ) throws Exception {
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java:2668]: Sure. Thanks!

File: model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java (hunk 1 above)
Comment (irinaepshteyn) [reply model/src/main/java/org/cloudfoundry/identity/uaa/zone/Links.java:101]: Added US554082 - [UAA] Apply split "self-service" flag to existing UAA zone configs

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 12]
```diff
@@ -531,33 +531,65 @@ private void loginThroughDiscovery(String userEmail, String password) {
     }
 
     @Test
-    public void testSelfServiceLinksBehavior() {
+    public void testSelfServiceResetPasswordLinksBehavior() {
         RestTemplate adminClient = IntegrationTestUtils.getClientCredentialsTemplate(
-                IntegrationTestUtils.getClientCredentialsResource(baseUrl, new String[0], "admin", "adminsecret"));
+            IntegrationTestUtils.getClientCredentialsResource(baseUrl, new String[0], "admin", "adminsecret"));
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:536]: Revert this change.  It's just noise

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 13]
```diff
@@ -531,33 +531,65 @@ private void loginThroughDiscovery(String userEmail, String password) {
...
         IntegrationTestUtils.createZoneOrUpdateSubdomain(adminClient, baseUrl, zoneId, zoneId, testZone3.getConfig());
+        webDriver.get(zoneUrl);
+        assertEquals(0, webDriver.findElements(By.xpath("//*[text()='Create account']")).size());
 
+        testZone3.getConfig().getLinks().setSelfService(new Links.SelfService().setSelfServiceCreateAccountEnabled(true).setSignup(null));
+        IntegrationTestUtils.createZoneOrUpdateSubdomain(adminClient, baseUrl, zoneId, zoneId, testZone3.getConfig());
+        webDriver.get(zoneUrl);
+        assertEquals(0, webDriver.findElements(By.xpath("//*[text()='Create account']")).size());
+
+        testZone3.getConfig().getLinks().setSelfService(new Links.SelfService().setSelfServiceCreateAccountEnabled(true).setSignup("http://example.com"));
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:580]: This test case is redundant; same test case is covered in lines 590-593

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 14]
```diff
@@ -531,33 +531,65 @@ private void loginThroughDiscovery(String userEmail, String password) {
...
+    public void testSelfServiceResetPasswordLinksBehavior() {
         RestTemplate adminClient = IntegrationTestUtils.getClientCredentialsTemplate(
-                IntegrationTestUtils.getClientCredentialsResource(baseUrl, new String[0], "admin", "adminsecret"));
+            IntegrationTestUtils.getClientCredentialsResource(baseUrl, new String[0], "admin", "adminsecret"));
         String zoneId = "testzone3";
         String zoneUrl = baseUrl.replace("localhost", zoneId+".localhost");
         IdentityZone testZone3 = IntegrationTestUtils.createZoneOrUpdateSubdomain(adminClient, baseUrl, zoneId, zoneId, new IdentityZoneConfiguration());
 
-        testZone3.getConfig().getLinks().setSelfService(new Links.SelfService().setSelfServiceLinksEnabled(true).setPasswd("").setSignup(""));
+        testZone3.getConfig().getLinks().setSelfService(new Links.SelfService().setSelfServiceResetPasswordEnabled(true).setPasswd(""));
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:541]: Minor.  Each test case below calls `new Links.SelfService()...`.  You could initialize it once at the beginning of the test and then re-use.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 15]
```diff
@@ -531,33 +531,65 @@ private void loginThroughDiscovery(String userEmail, String password) {
...
+
+    @Test
+    public void testSelfServiceCreateAccountLinksBehavior() {
+        RestTemplate adminClient = IntegrationTestUtils.getClientCredentialsTemplate(
+            IntegrationTestUtils.getClientCredentialsResource(baseUrl, new String[0], "admin", "adminsecret"));
+        String zoneId = "testzone3";
+        String zoneUrl = baseUrl.replace("localhost", zoneId+".localhost");
+        IdentityZone testZone3 = IntegrationTestUtils.createZoneOrUpdateSubdomain(adminClient, baseUrl, zoneId, zoneId, new IdentityZoneConfiguration());
+
+        testZone3.getConfig().getLinks().setSelfService(new Links.SelfService().setSelfServiceCreateAccountEnabled(true).setSignup(""));
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:570]: Minor. Each test case below calls `new Links.SelfService()...`. You could initialize it once at the beginning of the test and then re-use.

**Review by irinaepshteyn — APPROVED** [review]
@babamishra9, I approve it on principal.  There are a few minor things to clean up.  You could merge after it's done.

Please DO NOT merge into rc_75.18.1 branch yet as the feature is not completely implemented.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (irinaepshteyn) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/login/LoginMockMvcTests.java:2668]: Added US554095 - [UAA] Validate self-service links during login with IDP discovery

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java [hunk 16]
```diff
@@ -12,6 +12,7 @@
 import org.cloudfoundry.identity.uaa.util.UaaUrlUtils;
 import org.cloudfoundry.identity.uaa.zone.IdentityZoneHolder;
 import org.cloudfoundry.identity.uaa.zone.MergedZoneBrandingInformation;
+import org.cloudfoundry.identity.uaa.zone.Links.SelfService;
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply server/src/main/java/org/cloudfoundry/identity/uaa/account/ResetPasswordController.java]: Done.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 12 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java]: Done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 13 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java]: Done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 14 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java]: Done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 15 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java]: Done
